packages/plugin-node-surrounding-code/surrounding-code.js

```diff
@@ -28,16 +31,56 @@ module.exports = {
 }
 
 const getSurroundingCode = (file, lineNumber, cb) => {
-  readFile(file, 'utf8', (err, data) => {
-    if (err) return cb(err)
-    const lines = data.split('\n')
-    const start = Math.max(0, lineNumber - SURROUNDING_LINES - 1)
-    const end = Math.min(lines.length, lineNumber + SURROUNDING_LINES)
-    cb(null, lines
-      .slice(start, end)
-      .reduce((accum, line, i) => {
-        accum[start + (i + 1)] = line
-        return accum
-      }, {}))
+  const start = lineNumber - SURROUNDING_LINES
+  const end = lineNumber + SURROUNDING_LINES
+
+  const reader = createReadStream(file, { encoding: 'utf8' })
+  const splitter = new byline.LineStream({ keepEmptyLines: true })
+  const slicer = new CodeRange({ start, end })
+
+  // if the slicer has enough lines already, no need to keep reading from the file
+  slicer.on('done', () => reader.destroy())
```

Comment:
there's no way to cleanly kill a readable stream, you have to call `destroy()` on it which triggers an error. We tolerate this "premature close" error below. If we didn't do this, the entire file would be read chunk by chunk and just not operated on, which is a noop, but it does use resources for no reason.